tools/ami-cleanup/src/internal/cleanup.go

```diff
@@ -0,0 +1,239 @@
+ * limitations under the License.
+ */
+
+package internal
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/account"
+	accountTypes "github.com/aws/aws-sdk-go-v2/service/account/types"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2Types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/gravitational/trace"
+	"github.com/relvacode/iso8601"
+)
+
+type ApplicationInstance struct {
+	shouldDoDryRun bool
+
+	// Dependency injection for testing
+	accountClientGenerator func(cfg *aws.Config) IAccountApi
+	ec2ClientGenerator     func(cfg *aws.Config) IEc2Api
+}
+
+// Creates a new instance of the tool
+func NewApplicationInstance(doDryRun bool) *ApplicationInstance {
+	return &ApplicationInstance{
+		shouldDoDryRun:         doDryRun,
+		accountClientGenerator: NewAccountApi,
+		ec2ClientGenerator:     NewEc2Api,
+	}
+}
+
+// Performs cleanup
+func (ai *ApplicationInstance) Run(ctx context.Context) error {
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err, "failed to load AWS credentials")
+	}
+
+	enabledRegions, err := ai.getEnabledRegions(ctx, ai.accountClientGenerator(&cfg))
+	if err != nil {
+		return trace.Wrap(err, "failed to get enabled regions")
+	}
+
+	if len(enabledRegions) == 0 {
+		return nil
+	}
+
+	totalSpaceRecovered := int32(0)
+	totalImagesDeleted := 0
+	// This would probably run much faster with channels/concurrency, but there isn't much
+	// need with how often this is expected to be ran
+	for _, enabledRegion := range enabledRegions {
+		spaceRecovered, imagesDeleted, err := ai.cleanupRegion(ctx, cfg, *enabledRegion.RegionName)
+		if err != nil {
+			return trace.Wrap(err, "failed to clean up images in region %q", *enabledRegion.RegionName)
+		}
+
+		totalSpaceRecovered += spaceRecovered
+		totalImagesDeleted += imagesDeleted
+	}
+
+	log.Printf("Deleted %d GiB of %d images across %d regions", totalSpaceRecovered, totalImagesDeleted, len(enabledRegions))
+	return nil
+}
+
+// Cleans up the images in the given region. Returns the total amount of space recovered, number of images deleted,
+// and any error.
+func (ai ApplicationInstance) cleanupRegion(ctx context.Context, cfg aws.Config, regionName string) (int32, int, error) {
+	// A new client must be created for each region
+	cfg.Region = regionName
+	ec2Client := ai.ec2ClientGenerator(&cfg)
+
+	log.Printf("Deleting images in %q", regionName)
+	devImages, err := ai.getDevImagesInRegion(ctx, ec2Client)
+	if err != nil {
+		return 0, 0, trace.Wrap(err, "failed to get a list of dev images for %q", regionName)
+	}
+
+	var totalSpaceRecovered int32
+	totalImagesDeleted := 0
+	totalSnapshotCount := 0
+	for _, devImage := range devImages {
+		imageSpace, err := ai.cleanupImageIfOld(ctx, ec2Client, devImage)
+		if err != nil {
+			return 0, 0, trace.Wrap(err, "failed to cleanup dev image %q", devImage.Name)
```

Comment:
Should we continue if we failed to cleanup one image instead of returning on first failure?

Reply:
I think so - if one fails for whatever reason, then it is highly likely that the remaining will fail. The most common failures include invalid AWS credentials, API rate limits, bugs resulting in invalid queries.